### PR TITLE
Check in of generic reduction templates and some reductions

### PR DESCRIPTION
**Pull request by ndgrigorian**
This pull request implements the necessary meta-programming utilities for reductions with arbitrary binary operators, templates for generic reductions, as well as the new functions ``dpctl.tensor.max``, ``dpctl.tensor.min``, ``dpctl.tensor.argmax``, and ``dpctl.tensor.argmin``.

These meta-programming utilities, such as ``Identity``, make it possible to define unique aspects of a reduction such that it may be passed to a generic functor for computation.

- [X] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [X] If this PR is a work in progress, are you opening the PR as a draft?

### CODE REVIEW

File: dpctl/tensor/libtensor/include/utils/sycl_utils.hpp [hunk 1]
```diff
@@ -66,6 +131,172 @@ size_t choose_workgroup_size(const size_t nelems,
...
+    static constexpr std::complex<T> value{std::numeric_limits<T>::infinity(),
+                                           std::numeric_limits<T>::infinity()};
+};
+
+// Plus
+
+template <typename T, class Op>
+using IsPlus = std::bool_constant<
+    std::is_same_v<Op, sycl::plus<T>> || std::is_same_v<Op, sycl::plus<void>> ||
+    std::is_same_v<Op, std::plus<T>> || std::is_same_v<Op, std::plus<T>>>;
```
Comment (oleksandr-pavlyk) [inline dpctl/tensor/libtensor/include/utils/sycl_utils.hpp:280]:
```suggestion
template <typename T, class Op>
using IsPlus = std::bool_constant<
    std::is_same_v<Op, sycl::plus<T>> || std::is_same_v<Op, sycl::plus<void>> ||
    std::is_same_v<Op, std::plus<T>>>;
```

Duplicated condition. For the curiosity, why are we testing for `sycl::plus<void>`? Is it a blanket cover-all test for all types?

Comment (oleksandr-pavlyk) [reply dpctl/tensor/libtensor/include/utils/sycl_utils.hpp:280]: I see that the same test is being used in `sycl` headers. 🤷

Comment (ndgrigorian) [reply dpctl/tensor/libtensor/include/utils/sycl_utils.hpp:280]: It's done in the headers. I believe it's intended to test for the ``void`` specialization of the ``std`` functions (as the ``sycl`` variants are simply the function objects from the STL ``functional`` header), which deduces the return and parameter types (see: https://en.cppreference.com/w/cpp/utility/functional/plus_void)

These can be safely removed for our custom ops, which don't have those overloads, so I will go ahead and do that as well.

**Review by oleksandr-pavlyk — APPROVED** [review]
I think this PR is ready to go in. Subsequent development can be done in new PRs. Thank you @ndgrigorian !

**Pull request merged**